subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/DefaultResolutionResultBuilder.java

```diff
@@ -73,21 +73,25 @@ public void visitComponent(ResolvedGraphComponent component) {
     public void visitOutgoingEdges(Long fromComponent, Collection<? extends ResolvedGraphDependency> dependencies) {
         DefaultResolvedComponentResult from = modules.get(fromComponent);
         for (ResolvedGraphDependency d : dependencies) {
-            DependencyResult dependencyResult;
+            DependencyResult dependencyResult = null;
             ResolvedVariantResult fromVariant = d.getFromVariant();
             if (d.getFailure() != null) {
                 dependencyResult = dependencyResultFactory.createUnresolvedDependency(d.getRequested(), from, d.isConstraint(), d.getReason(), d.getFailure());
             } else {
                 DefaultResolvedComponentResult selected = modules.get(d.getSelected());
-                if (selected == null) {
+                if (selected == null && !d.isConstraint()) {
```

Comment:
This is not the right fix. The graph Must be correct once resolved. This is a symptom showing it is not the case.